packages/core/src/com/initializers/iframe.ts

```diff
@@ -34,47 +34,75 @@ export function iframeInitializer({
             : await useWindow(com, iframeElement, instanceId, src ?? defaultSourceFactory(env, publicPath));
 
         return {
-            id: instanceId,
+            id,
         };
     };
 }
-async function useWindow(com: Communication, host: WindowHost, instanceId: string, src: string): Promise<void> {
+async function useWindow(com: Communication, host: WindowHost, instanceId: string, src: string): Promise<string> {
     const win = isIframe(host) ? host.contentWindow : host;
     if (!win) {
         throw new Error('cannot spawn detached iframe.');
     }
     com.registerEnv(instanceId, win);
     await injectScript(win, instanceId, src);
     await com.envReady(instanceId);
+    return instanceId;
 }
 
-async function useIframe(com: Communication, host: HTMLIFrameElement, instanceId: string, src: string): Promise<void> {
+async function useIframe(
+    com: Communication,
+    host: HTMLIFrameElement,
+    instanceId: string,
+    src: string
+): Promise<string> {
     const win = host.contentWindow;
     if (!win) {
         throw new Error('cannot spawn detached iframe.');
     }
 
-    await changeLocation(win, host, instanceId, src);
-    com.registerEnv(instanceId, win);
-    await com.envReady(instanceId);
+    const previousWindowName = win.name;
 
-    const reload = () => com.reconnectHandlers(instanceId);
-    host.addEventListener('load', () => {
-        com.envReady(instanceId).then(reload).catch(reportError);
-    });
+    const locationChanged = await changeLocation(win, host, instanceId, src);
+    if (locationChanged) {
+        com.registerEnv(instanceId, win);
+        await com.envReady(instanceId);
+        const handleIframeReload = () => {
+            com.envReady(instanceId)
+                .then(() => com.reconnectHandlers(instanceId))
+                .catch(reportError);
+        };
+
+        host.addEventListener('load', handleIframeReload);
```

Comment:
But this will add another 'load' handler every time this function is called for the same iframe. Exactly what I talked about here: https://github.com/wixplosives/engine/pull/153#discussion_r413693010